src/dotnet/GeneratorAgent/src/Azure.Tools.GeneratorAgent/Program.cs

```diff
@@ -9,52 +10,130 @@ namespace Azure.Tools.GeneratorAgent
+            Logger = LoggerFactory.CreateLogger<Program>();
+            CommandLineConfig = new CommandLineConfiguration(LoggerFactory.CreateLogger<CommandLineConfiguration>());
+        }
+
+        public static async Task<int> Main(string[] args)
+        {
+            ToolConfiguration toolConfig = new();
+            using ILoggerFactory loggerFactory = toolConfig.CreateLoggerFactory();
+
+            Program program = new(toolConfig, loggerFactory);
+            return await program.RunAsync(args).ConfigureAwait(false);
+        }
+
+        internal async Task<int> RunAsync(string[] args)
         {
-            using CancellationTokenSource cts = new CancellationTokenSource();
+            RootCommand rootCommand = CommandLineConfig.CreateRootCommand(HandleCommandAsync);
+            return await rootCommand.InvokeAsync(args).ConfigureAwait(false);
+        }
+
+        internal async Task<int> HandleCommandAsync(string? typespecPath, string? commitId, string? typespecSpecDirectory, string sdkPath)
+        {
+            int validationResult = CommandLineConfig.ValidateInput(typespecPath, commitId, typespecSpecDirectory);
+            if (validationResult != ExitCodeSuccess)
+            {
+                return validationResult;
+            }
 
-            Console.CancelKeyPress += (sender, eventArgs) =>
+            using CancellationTokenSource cancellationTokenSource = new();
+            Console.CancelKeyPress += (_, eventArgs) =>
             {
-                cts.Cancel();
+                Logger.LogInformation("Cancellation requested by user");
+                cancellationTokenSource.Cancel();
                 eventArgs.Cancel = true;
             };
 
-            ToolConfiguration ToolConfig = new ToolConfiguration();
-            ILoggerFactory LoggerFactory = ToolConfig.CreateLoggerFactory();
-            AppSettings AppSettings = ToolConfig.CreateAppSettings();
-
-            ILogger<ErrorFixerAgent> AgentLogger = LoggerFactory.CreateLogger<ErrorFixerAgent>();
-            ILogger<CredentialFactory> CredentialLogger = LoggerFactory.CreateLogger<CredentialFactory>();
-            ILogger<Program> Logger = LoggerFactory.CreateLogger<Program>();
+            try
+            {
+                return await ExecuteGenerationAsync(typespecPath, commitId, typespecSpecDirectory, sdkPath, cancellationTokenSource.Token).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                Logger.LogInformation("Operation was cancelled. Shutting down gracefully");
+                return ExitCodeSuccess;
+            }
+            catch (Exception ex)
+            {
+                Logger.LogError(ex, "Unexpected error occurred during command execution");
+                return ExitCodeFailure;
+            }
+        }
 
+        private async Task<int> ExecuteGenerationAsync(
+            string? typespecPath,
```

Comment:
Each parameter here should be treated as untrusted.   Validate to the best of our ability.   The TypeSpec path should have an expected `.tsp` extension, for example.

Kicking off a process is very dangerous if we accept any user input related to it.   Be paranoid, and don't trust anything.   Better to get a bug because we're too restrictive than one that we're vulnerable to executing arbitrary commands.